src/origin-experiments.js

```diff
@@ -0,0 +1,172 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {bytesToString, stringToBytes} from './utils/bytes';
+import {getSourceOrigin, parseUrl} from './url';
+import {parseJson} from './json';
+
+/**
+ * Generates, signs and verifies origin experiments.
+ */
+export class OriginExperiments {
+  /**
+   * @param {!./service/crypto-impl.Crypto} crypto
+   */
+  constructor(crypto) {
+    /** @private {!./service/crypto-impl.Crypto} */
+    this.crypto_ = crypto;
+  }
+
+  /**
+   * Generates an RSA public/private key pair for signing and verifying.
+   * @return {!Promise}
+   */
+  generateKeys() {
+    const generationAlgo = Object.assign({
+      modulusLength: 2048,
+      publicExponent: Uint8Array.of(1, 0, 1),
+    }, this.crypto_.pkcsAlgo);
+    return this.crypto_.subtle.generateKey(
+        /** @type {{name: string}} */ (generationAlgo),
+        /* extractable */ true,
+        /* keyUsages */ ['sign', 'verify']);
+  }
+
+  /**
+   * Generates an origin experiment token given a config json.
+   * @param {number} version
+   * @param {!JsonObject} json
+   * @param {!webCrypto.CryptoKey} privateKey
+   * @return {!Promise<string>}
+   */
+  generateToken(version, json, privateKey) {
+    const config = stringToBytes(JSON.stringify(json));
+    const data = this.prepend_(version, config);
+    return this.sign_(data, privateKey).then(signature => {
+      return this.append_(data, new Uint8Array(signature));
+    });
+  }
+
+  /**
+   * Verifies an origin experiment token given a public key.
+   * @param {string} token
+   * @param {!Location} location
+   * @param {!webCrypto.CryptoKey} publicKey
+   * @return {!Promise<string>} If token is valid, resolves with the
+   *     experiment ID. Otherwise, rejects with validation error.
+   */
+  verifyToken(token, location, publicKey) {
+    return new Promise((resolve, reject) => {
+      let i = 0;
+      const bytes = stringToBytes(atob(token));
+
+      // Parse version.
+      const version = bytes[i];
+      if (version !== 0) {
+        reject(new Error(`Unrecognized token version: ${version}`));
+      }
+      i += 1;
+
+      // Parse config length.
+      const length = new DataView(bytes.buffer).getUint32(i);
+      i += 4; // Number of bytes in Uint32 config length.
+      if (length > bytes.length - i) {
+        reject(new Error(`Unexpected config length: ${length}`));
+      }
+
+      // Parse config itself.
+      const configBytes = bytes.subarray(i, i + length);
+      i += length;
+
+      // Parse unsigned data and its signature.
+      const data = bytes.subarray(0, i);
+      const signature = bytes.subarray(i);
+
+      this.verify_(signature, data, publicKey).then(verified => {
```

Comment:
You should be resolving with this expression, not waiting for an inner callback to call resolve.